explorer/interpreter/type_checker.cpp

```diff
@@ -912,24 +925,39 @@ auto TypeChecker::ArgumentDeduction(
 class ConstraintTypeBuilder {
  public:
   ConstraintTypeBuilder(Nonnull<Arena*> arena, SourceLocation source_loc)
-      : self_binding_(MakeSelfBinding(arena, source_loc)) {}
-  ConstraintTypeBuilder(Nonnull<const GenericBinding*> self_binding)
-      : self_binding_(self_binding) {}
+      : ConstraintTypeBuilder(arena, MakeSelfBinding(arena, source_loc)) {}
+  ConstraintTypeBuilder(Nonnull<Arena*> arena,
+                        Nonnull<GenericBinding*> self_binding)
+      : self_binding_(PrepareSelfBinding(arena, self_binding)),
+        impl_binding_(AddImplBinding(arena, self_binding_)) {}
 
   // Produce a type that refers to the `.Self` type of the constraint.
-  auto GetSelfType(Nonnull<Arena*> arena) const -> Nonnull<const Value*> {
+  auto GetSelfType() const -> Nonnull<const Value*> {
     return &self_binding_->value();
   }
 
+  // Get a witness that `.Self` implements the eventual constraint type built by
+  // this builder.
+  auto GetSelfWitness() const -> Nonnull<const Witness*> {
+    return cast<Witness>(impl_binding_->symbolic_identity().value());
+  }
+
+  // Get the impl binding for `.Self`.
+  auto GetImplBinding() const -> Nonnull<const ImplBinding*> {
```

Comment:
Hm, I don't think of these as properties of the builder -- I don't think I would write these as properties in Carbon -- but I'm not completely sure what our threshold is here. It would be reasonable for these functions to build a new value on demand instead of returning an existing one, which makes me lean towards non-accessor naming a little -- indeed, `GetSelfType` used to do that.

Happy to let you make the call here.

Reply:
GetImplBinding -> impl_binding at least seems straightforward to me because there's no hidden function call. If the behavior changes later, that may be a good time to change the name too. But I resist conservative naming on the idea that maybe it'll change, since maybe it won't change.

GetSelfType and GetSelfWitness I think is somewhat a matter of perspective -- I could see someone changing GetSelfType to self_type without objection, GetSelfWitness is slightly iffy due to the cast but still fits the intent I think. But I don't think these are as straightforward so I wouldn't insist on them.

Reply:
Turns out that `GetImplBinding` is unused, so I removed it.